src/utils/date.js

```diff
@@ -51,3 +53,49 @@ export function getDate(value) {
   // Must be a `Date` object.
   return value.getTime();
 }
+
+/** Map from attribute names to their parsers. */
+const dateAttrParsers = {
+  'datetime': (datetime) =>
+    userAssert(parseDate(datetime), `Invalid date: ${datetime}`),
+  'end-date': (datetime) =>
+    userAssert(parseDate(datetime), `Invalid date: ${datetime}`),
+  'timeleft-ms': (timeleftMs) => Date.now() + Number(timeleftMs),
+  'timestamp-ms': Number,
+  'timestamp-seconds': (timestampSeconds) => 1000 * Number(timestampSeconds),
+};
+
+/**
+ * @param {!Element} element
+ * @param {!Array<string>} dateAttrs list of attribute names
+ * @return {?number}
+ */
+export function parseDateAttrs(element, dateAttrs) {
+  const epoch = userAssert(
+    parseEpoch(element, dateAttrs),
+    `One of [${dateAttrs.join(', ')}] is required`
+  );
+
+  const offsetSeconds =
+    (Number(element.getAttribute('offset-seconds')) || 0) * 1000;
+  return epoch + offsetSeconds;
+}
+
+/**
+ * Parse epoch from list of possible element attributes, returning the first one
+ * that is truthy.
+ * @param {!Element} element
+ * @param {!Array<string>} dateAttrs list of attribute names
+ * @return {?number}
+ */
+function parseEpoch(element, dateAttrs) {
+  for (let i = 0; i < dateAttrs.length; ++i) {
+    const attrName = dateAttrs[i];
+    const attrVal = element.getAttribute(attrName);
+    if (attrVal) {
+      return dateAttrParsers[attrName](attrVal);
```

Comment:
Add a `devAssert` that `dateAttrParsers[attrName]` exists?

Reply:
Done